### ISSUE TITLE
Make mail address optional on signup

### ISSUE THREAD

**Issue by TiBiBa** [Felienne/hedy#2075]
Currently setting up a mail address is mandatory when creating an account. As they have no explicite added value it would be nice if we made this optional.

**Comment by TiBiBa** [comment]
I like to change my mind on this one. The mail address is essential for functionality such as password reset and account recovery. It might be an issue that students don't have mail addresses, but we should think about another work-around. For example: making mail addresses no longer unique but allow "whitelisted" (teacher) mail addresses to be used multiple times. 

One solution that we can offer teacher for the time being (one I use for testing on Alpha as well) is simply adding `+<number>` to the address. As these values are allowed they will be cut when the mail is sent, resulting in using the same mail address but from our back-end perspective it is unique. For example:

`pietjepuk@hotmail.com` and `pietjepuk+1@hotmail.com` are unique, but mails sent to both are always received by  `pietjepuk@hotmail.com`. What do you think @Felienne ?

**Comment by TiBiBa** [comment]
Will work on this today (and most likely tomorrow)! I do have an additional idea that I think we should implement to make sure accounts without mail addresses are still able to reset their password. We might want to return a "secret key" that users should save that can also be used to reset their password. In this case we have the following possible workflows:

1. User creates an account with a mail address
    -  They receive an verification mail address -> which also contains their username
    -  They are automatically logged in and directed to the landing-page

2. User creates an account without a mail address 
    - We store a "secret key" on the user account (This should be hashed!)
    - They are directed to a dedicated page -> explaining and showing this key
    - When pressing "accept" -> re-directed to the landing-page

3. On the password reset we verify if the user has a mail address:
    - If so -> Sent the normal "reset" mail as is the current flow
        - Sent "password is reset" mail
        - Re-direct to programs page 
    - If not -> Open dedicated page and let user enter secret key
        - Return user to dedicated page -> give a new key
        - Re-direct to programs page


I think this would be a nice solution where users without a mail address don't lose any functionality. But I do have some concerns: I'm not really familiar with hashing, security and the state-of-the-art Flask way-to-go around this, so will have to figure this out. Another concern: Anyone is able to find out if an account has a mail address or not, because this can be deducted from the workflow. Sidenote on this one: This is currently possible as well as we return a "user doesn't exist" error when trying to reset for a non-existing user".

I'm still struggling a bit with this issue and I think we have to think about this really well before starting implementation as we have to re-write some core-functionality to make sure this works. What are your thoughts on this? @Felienne @boryanagoncharenko

**Comment by boryanagoncharenko** [comment]
Apologies for the delayed response.

I like the idea of a generated secret key since it lets the user use all Hedy features without an email address (such as reset their password). My only concern is whether students would actually save this secret key (I know I will be lazy and won't save it). Obviously we could make it human friendly instead of a long hash, but this does not guarantee that students would save and use it later? If a student did not save their key and forgot their password (e.g. after a 2-week break), they will be forced to create a new account and re-join their class. This might be inconvenient for the teacher because the progress of the student will be split across 2 accounts.

Maybe this is a wild idea but could we allow email-less accounts to be created only in the context of a teacher. What I mean is a teacher can create email-less accounts and can reset their passwords and students can create email-less accounts as long as they supply the teacher's account (or the class they need to join) upon sign-up. In this way, if a student forgets their password, the teacher can reset it without getting duplicated student accounts in their class and keeping the progress of the student in a single account.

About your concern that anyone can find whether the account has an email or not and the side note regarding the current behaviour. From conversations with Felienne I learned that indeed we are cutting back on some of the standard practises intentionally. The rationale is that Hedy users are children that might need more context. Needless to say, this task is an excellent opportunity to reassess this approach.

**Comment by TiBiBa** [comment]
> Apologies for the delayed response.
> 
> I like the idea of a generated secret key since it lets the user use all Hedy features without an email address (such as reset their password). My only concern is whether students would actually save this secret key (I know I will be lazy and won't save it). Obviously we could make it human friendly instead of a long hash, but this does not guarantee that students would save and use it later? If a student did not save their key and forgot their password (e.g. after a 2-week break), they will be forced to create a new account and re-join their class. This might be inconvenient for the teacher because the progress of the student will be split across 2 accounts.
> 
> Maybe this is a wild idea but could we allow email-less accounts to be created only in the context of a teacher. What I mean is a teacher can create email-less accounts and can reset their passwords and students can create email-less accounts as long as they supply the teacher's account (or the class they need to join) upon sign-up. In this way, if a student forgets their password, the teacher can reset it without getting duplicated student accounts in their class and keeping the progress of the student in a single account.
> 
> About your concern that anyone can find whether the account has an email or not and the side note regarding the current behaviour. From conversations with Felienne I learned that indeed we are cutting back on some of the standard practises intentionally. The rationale is that Hedy users are children that might need more context. Needless to say, this task is an excellent opportunity to reassess this approach.

I agree! We already have functionality in place to enable teachers to create multiple accounts, we can (mostly) delete the related PR and only implement the email-less account creation there. However, we do have to keep the students linked to the teacher forever, out of scope of a class/student relation. We can use the following approach:

- Teacher can create email-less accounts -> only on the teacher specific "create multiple accounts" page
- These accounts are "linked" to the teacher with additional fields `student: True, teacher: ...`
- Teachers can retrieve an overview of their created accounts
- Teachers can already reset a password of a student in the class overview
   - We might want to move this to the "accounts created by teacher" view
- Teachers can alter all information on these accounts -> also remove them
- The teacher accounts can only remove their account if there are no related accounts without a mailaddress

**Comment by boryanagoncharenko** [comment]
Completely agree with your vision. The email-less accounts should be tied to the teacher (and not to a class) and the teacher should be able to manage them: create, modify or delete the account, reset password, list the existing accounts. I am not sure what to do with accounts that started email-less and then got an email. Probably when an account receives an email they should not be tied to the teacher any more? What do you think?

The only addition I could suggest is to let the students create an email-less account by providing the class they need to join. I could imagine the teacher would like to write down the name of the class students want to join and let them pick their usernames and passwords (opposed to the teacher having to create a username and password for every student). When the student specifies the course on sign-up, we will link the account to the teacher and also add the student to the class. Any thoughts on this suggestion?
